### PR TITLE
[AY1920S1-CS2113T-W17-3] OwlMoney

### PR DESCRIPTION
**Pull request by shamussy**
@shamussy
@briantkm
@terencetanwt
@mong-nus
@valerietyj

### CODE REVIEW

File: src/main/java/owlmoney/Main.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+    private Ui ui;
+    private ParseCommand parser;
+    private Profile profile;
+    //private Storage storage;
+
+    /**
+     * Initializes a new OwlMoney session.
+     */
+    private Main() {
```
Comment (jinyao-lee) [inline src/main/java/owlmoney/Main.java:36]: Method name should start with a verb (e.g.: `private initializeSession(...)`)

File: src/main/java/owlmoney/Main.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+import owlmoney.model.bond.exception.BondException;
+import owlmoney.model.goals.exception.GoalsException;
+import owlmoney.model.profile.Profile;
+import owlmoney.model.transaction.exception.TransactionException;
+import owlmoney.ui.Ui;
+
+/**
+ * The main class.
+ */
+
```
Comment (jinyao-lee) [inline src/main/java/owlmoney/Main.java]: Should not have a blank line here

File: src/main/java/owlmoney/Main.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    private Main() {
+        ui = new Ui();
+        parser = new ParseCommand();
+        /*storage = new Storage("data/data.txt");
+        try {
+            tasks = new TaskList(storage.readFile());
+        } catch (FileNotFoundException e) {
+            ui.printError("Could not read tasks from disk, will start with empty file");
+            tasks = new TaskList();
+        }*/
```
Comment (jinyao-lee) [inline src/main/java/owlmoney/Main.java]: Commented code does not help the readability and should be deleted

File: src/main/java/owlmoney/Main.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        ui.greet(profile.profileGetUsername());
+        // until above this line
+        while (parser.hasNextLine()) {
+            try {
+                Command command = parser.parseLine();
+                hasExited = command.execute(profile, ui);
+                if (hasExited) {
+                    break;
+                }
+
```
Comment (jinyao-lee) [inline src/main/java/owlmoney/Main.java]: Unnecessary blank line

File: src/main/java/owlmoney/logic/command/Command.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    protected boolean isExit = false;
+
+    /**
+     * Abstract method where each command type implements execution code.
+     *
+     * @param profile Profile of the user.
+     * @param ui      Ui of OwlMoney.
+     * @return True if OwlMoney should terminate after execution.
+     */
+    public abstract boolean execute(Profile profile, Ui ui) throws BankException,
```
Comment (jinyao-lee) [inline src/main/java/owlmoney/logic/command/Command.java]: Docs comments for methods should start with a verb

File: src/main/java/owlmoney/logic/command/PlaceHolderEmptyCommand.java [hunk 6]
```diff
@@ -0,0 +1,17 @@
+package owlmoney.logic.command;
+
+import owlmoney.model.profile.Profile;
+import owlmoney.ui.Ui;
+
+public class PlaceHolderEmptyCommand extends Command {
```
Comment (jinyao-lee) [inline src/main/java/owlmoney/logic/command/PlaceHolderEmptyCommand.java]: "What's the purpose of this class?" --> Each class should have a Javadocs comment to explain the purpose of this class

File: src/main/java/owlmoney/logic/command/bank/DeleteInvestmentCommand.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+import owlmoney.model.bank.exception.BankException;
+import owlmoney.model.profile.Profile;
+import owlmoney.ui.Ui;
+
+/**
+ * DeleteInvestmentCommand contains the execution function to delete an investment object.
+ */
+public class DeleteInvestmentCommand extends Command {
+    private final String bankName;
+    private static final String INVESTMENT = "investment";
```
Comment (jinyao-lee) [inline src/main/java/owlmoney/logic/command/bank/DeleteInvestmentCommand.java]: `INVESTMENT = "investment"` doesn't explain the purpose of this variable. Is this used as a bank account type? If so, rename it such that readers will know that you are trying to convey that this is an account type

File: src/main/java/owlmoney/logic/parser/ParseCommand.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+/**
+ * Represents the first instance of parsing user input.
+ * This determines what type of command the user has entered.
+ */
+
+public class ParseCommand extends Parser {
+    private ParseType parseType = new ParseType();
+    private final Scanner scanner = new Scanner(System.in);
+    private static final String[] COMMAND_KEYWORDS = new String[] {
```
Comment (jinyao-lee) [inline src/main/java/owlmoney/logic/parser/ParseCommand.java:33]: The contents of your `COMMAND_KEYWORDS` are equivalent to **magic numbers**.

File: src/main/java/owlmoney/logic/parser/ParseCommand.java [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+        case "/add":
+            // Fallthrough
+        case "/delete":
+            // Fallthrough
+        case "/edit":
+            // Fallthrough
+        case "/list":
+            return parseType.parseData(command, data);
+        case "/exit":
+            return new ExitCommand();
```
Comment (jinyao-lee) [inline src/main/java/owlmoney/logic/parser/ParseCommand.java]: Magic numbers

File: src/main/java/owlmoney/logic/parser/ParseRawData.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
+package owlmoney.logic.parser;
+
+import owlmoney.logic.parser.exception.ParserException;
+
+/**
+ * Represents the third layer of parsing for raw data after removing command and type.
+ * This forms the baseline raw data parsing that specific data level parsers can extend from.
+ */
+
```
Comment (jinyao-lee) [inline src/main/java/owlmoney/logic/parser/ParseRawData.java]: Unnecessary blank line. The javadoc comments should be directly above your class or method without any blank lines

File: src/main/java/owlmoney/logic/parser/ParseRawData.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+        return parameter;
+    }
+
+    private void checkDuplicateKeywords(String[] splitArray, String[] keywordList) throws ParserException {
+        for (int i = 0; i < keywordList.length; i++) {
+            checkEachKeyword(keywordList[i], splitArray);
+        }
+    }
+
+    private void checkEachKeyword(String keyword, String[] splitArray) throws ParserException {
```
Comment (jinyao-lee) [inline src/main/java/owlmoney/logic/parser/ParseRawData.java:55]: What is checkEachKeyword?

File: src/main/java/owlmoney/logic/parser/ParseRawData.java [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+        int startIndex = findStartIndex(splitTypeSpecificArguments, keyword);
+        if (startIndex >= 0) {
+            int endIndex = findEndIndex(startIndex, splitTypeSpecificArguments, keywordList);
+            parameter = concatenateParameter(startIndex, endIndex, splitTypeSpecificArguments);
+        }
+        return parameter;
+    }
+
+    private void checkDuplicateKeywords(String[] splitArray, String[] keywordList) throws ParserException {
```
Comment (jinyao-lee) [inline src/main/java/owlmoney/logic/parser/ParseRawData.java:42]: The method name trips the reader (tripped me) as it is vague. Use a more relevant name like `assertNoDuplicateKeywords`. 

A note in general (may/may not apply to here): you should provide javadoc comments for private methods as well if they are non-trivial.

File: src/main/java/owlmoney/logic/parser/ParseRawData.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+        }
+    }
+
+    private int findStartIndex(String[] splitTypeSpecificArguments, String keyword) {
+        for (int i = 0; i < splitTypeSpecificArguments.length; i++) {
+            if (splitTypeSpecificArguments[i].equals(keyword)) {
+                return i;
+            }
+        }
+        return -1;
```
Comment (jinyao-lee) [inline src/main/java/owlmoney/logic/parser/ParseRawData.java]: Magic numbers

File: src/main/java/owlmoney/logic/parser/ParseType.java [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+        return parseTypeMenu(command, type, rawData);
+    }
+
+    /**
+     * Checks if the user wants to delete profile.
+     *
+     * @param command The extracted first field from the initial user input that determines the command.
+     * @throws ParserException if the user wants to delete his profile.
+     */
+    private void isDeleteProfile(String command) throws ParserException {
```
Comment (jinyao-lee) [inline src/main/java/owlmoney/logic/parser/ParseType.java]: `is` keyword should only be used if the method returns a boolean.

Use keywords such as assert or validate.

File: src/main/java/owlmoney/logic/parser/ParseType.java [hunk 15]
```diff
@@ -0,0 +1,291 @@
...
+                editGoals.checkParameter();
+                return editGoals.getCommand();
+            } else if ("/list".equals(command)) {
+                return new ListGoalsCommand();
+            }
+            throw new ParserException("You entered an invalid type for goals");
+        default:
+            throw new ParserException("You entered an invalid type");
+        }
+    }
```
Comment (jinyao-lee) [inline src/main/java/owlmoney/logic/parser/ParseType.java]: This method is wayyy too lengthy... Magic numbers as well. Find a way to shorten this!

image

File: src/main/java/owlmoney/logic/regex/RegexUtil.java [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+    /**
+     * Checks whether the interest rate entered is within the 100% limit.
+     * The first line of checks determines whether it is a 2 digit number up to 2 decimal places.
+     * The second line of checks determines whether it is less than 100%.
+     *
+     * @param input The user input that is subject to Regex checking.
+     * @return the result of the check on whether it fulfills the criteria.
+     */
+    public static boolean regexCheckInterestRate(String input) {
+        final String interestRegex = "^\\s*(?=.*[1-9])\\d{1,2}(\\.\\d{1,2})?$";
```
Comment (jinyao-lee) [inline src/main/java/owlmoney/logic/regex/RegexUtil.java]: Line 56 (seems) to have the same regex as well. Make it a class level variable instead.

I.e.: avoid duplicate codes

File: src/main/java/owlmoney/model/bank/BankList.java [hunk 17]
```diff
@@ -0,0 +1,589 @@
...
+     * @param amount   New amount of bank account.
+     * @param ui       required for printing.
+     * @throws BankException If duplicate bank name found.
+     */
+    public void bankListEditInvestment(String bankName, String newName, String amount, Ui ui)
+            throws BankException {
+        for (int i = 0; i < bankLists.size(); i++) {
+            if (bankLists.get(i).getAccountName().equals(bankName)
+                    && "investment".equals(bankLists.get(i).getType())) {
+                if (!(newName.isEmpty() || newName.isBlank())) {
```
Comment (jinyao-lee) [inline src/main/java/owlmoney/model/bank/BankList.java]: `asd.isBlank()` is quivalent to `asd.trim().isempty()`.

You can eliminate the use of isEmpty in this case. same as for those above

File: src/main/java/owlmoney/model/transaction/TransactionList.java [hunk 18]
```diff
@@ -0,0 +1,315 @@
...
+     * @param expNum   Transaction number of the expenditure.
+     * @param desc     New description of the expenditure.
+     * @param amount   New amount of the expenditure.
+     * @param date     New date of the expenditure.
+     * @param category New category of the expenditure.
+     * @param ui       required for printing.
+     * @return New amount of the expenditure.
+     * @throws TransactionException If incorrect date format.
+     */
+    public double editEx(int expNum, String desc, String amount, String date, String category, Ui ui)
```
Comment (jinyao-lee) [inline src/main/java/owlmoney/model/transaction/TransactionList.java]: Avoid such shorthands for your method name. This trips the readers

**Review by jinyao-lee — COMMENTED** [review]
Overall good job (with the help of Travis weeding out other checkstyle issues).

Note that the comments are not exhaustive (whatever I commented also applies to those below/above).

Things to note (the more serious issues):
- Avoid long method (one of your method is > 100 lines long!). Readers will have a hard time reading the code and you will have a tough time debugging that method.
- Avoid magic numbers. This problem occurred in many parts of your code.

**Comment by shamussy** [comment]
Hi Jin Yao,

With regards to your:

The contents of your COMMAND_KEYWORDS are equivalent to magic numbers.

I thought we did the right thing by putting all the supposedly magic strings in a declared constant with private static final?

Regards
Shamus Neo
________________________________

**Comment by shamussy** [comment]
Hi @jylee-git ,

With regards to your:

The contents of your COMMAND_KEYWORDS are equivalent to magic numbers.

I thought we did the right thing by putting all the supposedly magic strings in a declared constant with private static final?

Regards
Shamus Neo

File: src/main/java/owlmoney/logic/parser/ParseCommand.java (hunk 8 above)
Comment (jinyao-lee) [reply src/main/java/owlmoney/logic/parser/ParseCommand.java:33]:
```
Hi @jylee-git ,

With regards to your:

The contents of your COMMAND_KEYWORDS are equivalent to magic numbers.

I thought we did the right thing by putting all the supposedly magic strings in a declared constant with private static final?

Regards
Shamus Neo
```

Maybe let me rephrase it: It's equivalent to dumping a collection of magic numbers inside.

Because just dumping these `"/help", "/undo", "/add", "/edit", "/delete", "/list", "/find", "/transfer", "/exit"` don't explain what are these values about.

So for this sort of case, you should declare static final variables for each these commands, then dump the created static final variables in to your `private static final List<String> COMMAND_KEYWORD_LISTS`.

Something like this:

```java
private static final String CMD_HELP = "/help";
//...
private static final List<String> CMD_KEYWRDS = Arrays.asList(CMD_HELP, CMD_ADD, CMD_otherCmdsYouWantToPut);
```

**Comment by jinyao-lee** [comment]
P.S.: If anything, comment directly on the ones that I've commented

image

File: src/main/java/owlmoney/logic/parser/ParseCommand.java (hunk 8 above)
Comment (shamussy) [reply src/main/java/owlmoney/logic/parser/ParseCommand.java:33]: Alright noted, will correct them

File: src/main/java/owlmoney/Main.java (hunk 1 above)
Comment (shamussy) [reply src/main/java/owlmoney/Main.java]: Fixed in PR 108, I am not using hashtags to prevent tagging other groups.

File: src/main/java/owlmoney/Main.java (hunk 2 above)
Comment (shamussy) [reply src/main/java/owlmoney/Main.java]: Fixed in PR 108, I am not using hashtags to prevent tagging other groups.

File: src/main/java/owlmoney/Main.java (hunk 4 above)
Comment (shamussy) [reply src/main/java/owlmoney/Main.java]: Fixed in PR 108, I am not using hashtags to prevent tagging other groups.

File: src/main/java/owlmoney/logic/command/Command.java (hunk 5 above)
Comment (shamussy) [reply src/main/java/owlmoney/logic/command/Command.java]: Fixed in PR 110 and 111, I am not using hashtags to prevent tagging other groups.

File: src/main/java/owlmoney/logic/parser/ParseRawData.java (hunk 10 above)
Comment (shamussy) [reply src/main/java/owlmoney/logic/parser/ParseRawData.java]: Fixed in PR 108, 110 and 111, I am not using hashtags to prevent tagging other groups.

File: src/main/java/owlmoney/Main.java (hunk 3 above)
Comment (shamussy) [reply src/main/java/owlmoney/Main.java]: Fixed in PR 130, these codes were temporary placeholders before  I implemented persistent storage for OwlMoney.

File: src/main/java/owlmoney/logic/command/PlaceHolderEmptyCommand.java (hunk 6 above)
Comment (shamussy) [reply src/main/java/owlmoney/logic/command/PlaceHolderEmptyCommand.java]: Fixed in PR 131, was a placeholder to prevent errors from always appearing since we have not implemented the command previously

File: src/main/java/owlmoney/logic/regex/RegexUtil.java (hunk 16 above)
Comment (shamussy) [reply src/main/java/owlmoney/logic/regex/RegexUtil.java]: Fixed in PR 210

File: src/main/java/owlmoney/logic/command/bank/DeleteInvestmentCommand.java (hunk 7 above)
Comment (shamussy) [reply src/main/java/owlmoney/logic/command/bank/DeleteInvestmentCommand.java]: Fixed in PR 210

File: src/main/java/owlmoney/logic/parser/ParseCommand.java (hunk 9 above)
Comment (shamussy) [reply src/main/java/owlmoney/logic/parser/ParseCommand.java]: Fixed in PR 210

File: src/main/java/owlmoney/model/transaction/TransactionList.java (hunk 18 above)
Comment (shamussy) [reply src/main/java/owlmoney/model/transaction/TransactionList.java]: Fixed before PR 210

File: src/main/java/owlmoney/model/bank/BankList.java (hunk 17 above)
Comment (shamussy) [reply src/main/java/owlmoney/model/bank/BankList.java]: Fixed in PR 218

File: src/main/java/owlmoney/logic/parser/ParseType.java (hunk 15 above)
Comment (shamussy) [reply src/main/java/owlmoney/logic/parser/ParseType.java]: Fixed in PR 222

File: src/main/java/owlmoney/logic/parser/ParseRawData.java (hunk 11 above)
Comment (shamussy) [reply src/main/java/owlmoney/logic/parser/ParseRawData.java:55]: Resolved in PR 225

File: src/main/java/owlmoney/logic/parser/ParseRawData.java (hunk 12 above)
Comment (shamussy) [reply src/main/java/owlmoney/logic/parser/ParseRawData.java:42]: Resolved in PR 225

File: src/main/java/owlmoney/logic/parser/ParseRawData.java (hunk 13 above)
Comment (shamussy) [reply src/main/java/owlmoney/logic/parser/ParseRawData.java]: Resolved in PR 225

File: src/main/java/owlmoney/logic/parser/ParseType.java (hunk 14 above)
Comment (shamussy) [reply src/main/java/owlmoney/logic/parser/ParseType.java]: Resolved in PR 225
